src/redis/windowed_cache_impl.go

```diff
@@ -0,0 +1,263 @@
+package redis
+
+import (
+	"math"
+	"math/rand"
+
+	"github.com/coocood/freecache"
+	pb "github.com/envoyproxy/go-control-plane/envoy/service/ratelimit/v3"
+	"github.com/envoyproxy/ratelimit/src/assert"
+	"github.com/envoyproxy/ratelimit/src/config"
+	"github.com/envoyproxy/ratelimit/src/limiter"
+	"github.com/envoyproxy/ratelimit/src/utils"
+	"github.com/golang/protobuf/ptypes/duration"
+	logger "github.com/sirupsen/logrus"
+	"golang.org/x/net/context"
+)
+
+type windowedRateLimitCacheImpl struct {
+	client Client
+	// Optional Client for a dedicated cache of per second limits.
+	// If this client is nil, then the Cache will use the client for all
+	// limits regardless of unit. If this client is not nil, then it
+	// is used for limits that have a SECOND unit.
+	perSecondClient            Client
+	timeSource                 limiter.TimeSource
+	jitterRand                 *rand.Rand
+	expirationJitterMaxSeconds int64
+	cacheKeyGenerator          limiter.CacheKeyGenerator
+	localCache                 *freecache.Cache
+	nearLimitRatio             float32
+}
+
+func maxInt64(a int64, b int64) int64 {
+	if a > b {
+		return a
+	}
+	return b
+}
+
+func minInt64(a int64, b int64) int64 {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+func nanosecondsToDuration(nanoseconds int64) *duration.Duration {
+	nanos := nanoseconds
+	secs := nanos / 1e9
+	nanos -= secs * 1e9
+	return &duration.Duration{Seconds: secs, Nanos: int32(nanos)}
+}
+
+func secondsToNanoseconds(second int64) int64 {
+	return second * 1e9
+}
+
+func nanosecondsToSeconds(nanoseconds int64) int64 {
+	return nanoseconds / 1e9
+}
```

Comment:
Are there no helpers in the Go proto code that already do this? There are for C++.

Reply:
Cannot found it on proto duration
Should I move these code to utils too?